File: typescript/packages/core/src/core/github/readdir.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GitHubAccessor } from '../../accessor/github.ts'
import { LookupStatus } from '../../cache/index/config.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import type { PathSpec } from '../../types.ts'
import { fetchDirTree } from './_client.ts'
import { indexEntryForChild } from './entry.ts'
import { stripSlash } from '../../util/slash.ts'

function enoent(path: string): Error {
  const e = new Error(`ENOENT: ${path}`) as Error & { code: string }
  e.code = 'ENOENT'
  return e
}

function stripPrefix(path: PathSpec): string {
  const prefix = path.prefix
  let p = path.pattern !== null ? path.directory : path.original
  if (prefix !== '' && p.startsWith(prefix)) {
    p = p.slice(prefix.length) || '/'
  }
  return p
}

function normalizeKey(p: string): string {
  const trimmed = stripSlash(p)
  return trimmed === '' ? '/' : `/${trimmed}`
}

export async function readdir(
  accessor: GitHubAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<string[]> {
  if (index === undefined) {
    throw enoent(path.original)
  }
  const prefix = path.prefix
  const stripped = stripPrefix(path)
  const key = normalizeKey(stripped)

  const listing = await index.listDir(key)
  if (listing.entries !== undefined && listing.entries !== null) {
    return prefix !== '' && listing.entries.length > 0 && !listing.entries[0]?.startsWith(prefix)
      ? listing.entries.map((e) => prefix + e)
      : listing.entries
  }
  if (listing.status === LookupStatus.NOT_FOUND) {
    if (accessor.truncated) {
      return fallbackReaddir(accessor, key, index, prefix)
    }
    throw enoent(stripped)
  }
  return []
}

async function fallbackReaddir(
  accessor: GitHubAccessor,
  key: string,
  index: IndexCacheStore,
  prefix: string,
): Promise<string[]> {
  const parentSha = await resolveDirSha(accessor, key, index)
  if (parentSha === null) throw enoent(key)
  const entries = await fetchDirTree(accessor.transport, accessor.owner, accessor.repo, parentSha)
  const childKeys: string[] = []
  const childEntries: [string, ReturnType<typeof indexEntryForChild>][] = []
  for (const e of entries) {
    const childKey = `${key === '/' ? '' : key}/${e.path}`
    childKeys.push(childKey)
    childEntries.push([childKey, indexEntryForChild(e.path, e.sha, e.type, e.size ?? null)])
  }
  childKeys.sort()
  await index.setDir(key, childEntries)
  return childKeys.map((k) => (prefix !== '' ? prefix + k : k))
}

async function resolveDirSha(
  accessor: GitHubAccessor,
  key: string,
  index: IndexCacheStore,
): Promise<string | null> {
  const result = await index.get(key)
  if (result.entry !== undefined && result.entry !== null) {
    return result.entry.id
  }
  const parts = stripSlash(key)
    .split('/')
    .filter((p) => p !== '')
  let currentSha = accessor.ref
  let currentPath = ''
  for (const part of parts) {
    const entries = await fetchDirTree(
      accessor.transport,
      accessor.owner,
      accessor.repo,
      currentSha,
    )
    const found = entries.find((e) => e.path === part)
    if (found === undefined) return null
    currentSha = found.sha
    currentPath += `/${part}`
    await index.put(
      currentPath,
      indexEntryForChild(part, found.sha, found.type, found.size ?? null),
    )
  }
  return currentSha
}

File: python/mirage/core/github/readdir.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import logging

from mirage.cache.index import IndexCacheStore, IndexEntry, LookupStatus
from mirage.core.github.tree import fetch_dir_tree
from mirage.types import PathSpec
from mirage.utils.errors import enoent

log = logging.getLogger(__name__)


async def readdir(accessor, path: PathSpec,
                  index: IndexCacheStore) -> list[str]:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    virtual = path.original
    if isinstance(path, PathSpec):
        prefix = path.prefix
        path = path.directory if path.pattern else path.original
    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    path = path.rstrip("/") or "/"
    listing = await index.list_dir(path)
    if listing.entries is not None:
        if prefix and listing.entries and not listing.entries[0].startswith(
                prefix):
            return [prefix + e for e in listing.entries]
        return listing.entries
    if listing.status == LookupStatus.NOT_FOUND:
        if accessor.truncated:
            return await _fallback_readdir(accessor, path, index, virtual,
                                           prefix)
        raise enoent(virtual)
    return []


async def _fallback_readdir(
    accessor,
    path: str,
    index: IndexCacheStore,
    virtual: str,
    prefix: str = "",
) -> list[str]:
    """Per-directory tree fetch when recursive tree was truncated."""
    parent_sha = await _resolve_dir_sha(accessor, path, index)
    if parent_sha is None:
        raise enoent(virtual)
    entries = await fetch_dir_tree(accessor.config, accessor.owner,
                                   accessor.repo, parent_sha)
    norm = "/" + path.strip("/")
    child_keys: list[str] = []
    for entry in entries:
        child_path = norm + "/" + entry.path
        resource_type = "folder" if entry.type == "tree" else "file"
        idx_entry = IndexEntry(
            id=entry.sha,
            name=entry.path,
            resource_type=resource_type,
            size=entry.size,
        )
        index._entries[child_path] = idx_entry
        child_keys.append(child_path)
    index._children[norm] = sorted(child_keys)
    log.debug("fallback readdir populated %d entries for %s", len(entries),
              path)
    virtual_keys = sorted((prefix + k if prefix else k) for k in child_keys)
    return virtual_keys


async def _resolve_dir_sha(accessor, path: str,
                           index: IndexCacheStore) -> str | None:
    """Get the tree SHA for a directory path.

    Walks from root if needed, fetching per-directory trees.
    """
    norm = "/" + path.strip("/")
    result = await index.get(norm)
    if result.entry is not None:
        return result.entry.id
    parts = norm.strip("/").split("/")
    current_sha = accessor.ref
    current_path = ""
    for part in parts:
        entries = await fetch_dir_tree(accessor.config, accessor.owner,
                                       accessor.repo, current_sha)
        found = False
        for entry in entries:
            if entry.path == part:
                current_sha = entry.sha
                current_path += "/" + part
                idx_entry = IndexEntry(
                    id=entry.sha,
                    name=entry.path,
                    resource_type="folder" if entry.type == "tree" else "file",
                    size=entry.size,
                )
                index._entries[current_path] = idx_entry
                found = True
                break
        if not found:
            return None
    return current_sha
